File: NASA.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:dio/dio.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_stripe/flutter_stripe.dart';

class Nasa{
  static const My_KEY ="pXI0cr8DdRgEmMHuWQz8WxitLjegXkSW9HSZldyx";
  static const NASA_1000thMarssol = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?sol=1000&api_key=pXI0cr8DdRgEmMHuWQz8WxitLjegXkSW9HSZldyx";
  static const NASA_Earth = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?earth_date=";
  static const NASA_1000_Opportunity = "https://api.nasa.gov/mars-photos/api/v1/rovers/opportunity/photos?sol=1000&api_key=";
  static const NASA_Spirit = "https://api.nasa.gov/mars-photos/api/v1/rovers/spirit/photos?sol=1000&api_key=";
  static const NASA_1000cameras = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?sol=1000&camera=";
  String date = "";
  var mars =[""];
  var earth =[""];
  var resultcomp =[""];
  var cameracur =[""];
  var cameras  =["FHAZ","RHAZ","MASTCHEMCAM", "MAHLI", "MARDI",	"NAVCAM","PANCAM","MINITES"];
  var resultvalidation  = [""];
  var validation ;
  var ascending =[""];
  var descending =[""];
  int a;
  int b;





  //Get the 10 picst from the URL 1000 Mars sol
  Future<String>Query10pics1000Marssol() async{
    try{
      final response =await http.get(
          Uri.parse(NASA_1000thMarssol),

      ).then((response)async{
        String Marsphotos = jsonDecode(response.body).slice(0,10);

        return  Marsphotos;
      }).catchError((err){
        return 'Error $err';
      });
    }catch(e){return null;}
  }
///////////////////////////////////////////////////////////////Get the Earth date //////////////////////////////////////////////////////////

  Future<String>QueryEARTHDate() async{
    try{
      final response =await http.get(
        Uri.parse(NASA_1000thMarssol),

      ).then((response)async{
        String date = jsonDecode(response.body)['earth_date'];
        return date;
      }).catchError((err){
        return 'Error $err';
      });
    }catch(e){return null;}

  }
///////////////////////////////////////////////////////////////Get the pictures from the earth date//////////////////////////////////////////////////////////////////
  Future<String>QueryEARTHDate2() async{
    try{
      final response =await http.get(
        Uri.parse(NASA_Earth + date + My_KEY),

      ).then((response)async{
        String Earthphotos = jsonDecode(response.body).slice(0,10);
        return Earthphotos;
      }).catchError((err){
        return 'Error $err';
      });
    }catch(e){return null;}

  }
////////////////////////////////////////////////////////////////Comparing 1000Mars against Earth date picture///////////////////////////////////////////////////////////
  Future<String>Comparisons() async {
    for (int i = 0; i < 10; i++){
/////////////////////////////Getting the pictures from 1000Mars sol//////////////////////////////////
      try {
        final response = await http.get(
          Uri.parse(NASA_1000thMarssol),

        ).then((response) async {
          /////////////////////////////////////Storing data on a Matrix//////////////////////////////////////
           mars[i] = jsonDecode(response.body).slice(0,i);
          return date;
        }).catchError((err) {
          return 'Error $err';
        });
      } catch (e) {
        return null;
      }
      /////////////////////////////Getting the pictures from Earth date//////////////////////////////////
      try {
        final response = await http.get(
          Uri.parse(NASA_Earth + date + My_KEY),

        ).then((response) async {
          /////////////////////////////////////Storing data on a Matrix//////////////////////////////////////
          earth[i] = jsonDecode(response.body).slice(0,i);
          return date;
        }).catchError((err) {
          return 'Error $err';
        });
      } catch (e) {
        return null;
      }
////////////////////////////////////Making the comparison //////////////////////////////////////////////
      if (mars[i] == earth[i])
        {
          resultcomp[i]="Yes";
        }
      else {resultcomp[i]="Not the same"; }

  }
  }
  /////////////////////////////////////Validation no camera takes more than 10 times pictures than other/////////////////////////////////////////////////
  Future<String>Validation() async {
    for (int i = 0; i < 8; i++) {
      try {
        final response = await http.get(
          /////////////////////////////////////////calling the data for 1000 Mars on each camera and storing the data on a matrix
          Uri.parse(NASA_1000cameras + cameras[i] + My_KEY),

        ).then((response) async {

          cameracur [i] = jsonDecode(response.body);
          return cameracur [i];
        }).catchError((err) {
          return 'Error $err';
        });
      } catch (e) {
        return null;

      }
      //////////////////////////////////////recoding the data to a Json to get the length///////////////////////
      resultvalidation[i] = jsonDecode(cameracur[i]).length;
      /////////////////////////////////////sorting the data to descending order////////////////////////////////
      ascending = resultvalidation..sort();
      descending = ascending.reversed;

      ///////////////////////////////////diving by between the numbers (the first one is the highest so it will not change)/////////////////////////////
      for (int i = 0; i < 7; i++) {
        a = (descending[0] as int);
        b = (descending[i+1] as int);
        ///////////////////////////////////////the result must be less than 10 /////////////////////////////////////////////////////////////////////
        if (a/b>= 10 ){print ("There is a camera taking more than 10 times picture than other");}
          else {print ("No issues found");}



      }
      }

     }








}

File: NASA.cpp
#include <iostream>
#include "curl/curl.h"
#include "json/json.h"
#include <string>
#include <cstdint>
#include <memory>

#pragma comment (lib, "Normaliz.lib")
#pragma comment (lib, "Ws2_32.lib")
#pragma comment (lib, "Wldap32.lib")
#pragma comment (lib, "advapi32.lib")
#pragma comment (lib, "crypt32.lib")

namespace
{
    std::size_t callback(
        const char* in,
        std::size_t size,
        std::size_t num,
        std::string* out)
    {
        const std::size_t totalBytes(size * num);
        out->append(in, totalBytes);
        return totalBytes;
    }
}
const std::string url = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?sol=1000&api_key=pXI0cr8DdRgEmMHuWQz8WxitLjegXkSW9HSZldyx";
const std::string My_KEY = "pXI0cr8DdRgEmMHuWQz8WxitLjegXkSW9HSZldyx";
const std::string NASA_1000thMarssol = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?sol=1000&api_key=pXI0cr8DdRgEmMHuWQz8WxitLjegXkSW9HSZldyx";
const std::string NASA_Earth = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?earth_date=";
const std::string NASA_1000_Opportunity = "https://api.nasa.gov/mars-photos/api/v1/rovers/opportunity/photos?sol=1000&api_key=";
const std::string NASA_Spirit = "https://api.nasa.gov/mars-photos/api/v1/rovers/spirit/photos?sol=1000&api_key=";
const std::string NASA_1000cameras = "https://api.nasa.gov/mars-photos/api/v1/rovers/curiosity/photos?sol=1000&camera=";
 

int main()
{

  //  auto etriesArray = {""};
    auto earthday = "";
    int i = 0;
    std::string earthdate = "2015-05-30";

    CURL* curl;


    curl = curl_easy_init();

    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    
    //Timeout 10sec
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10);
    
    // Response information.
    long httpCode(0);
    std::unique_ptr<std::string> httpData(new std::string());
    // Hook up data handling function.
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, callback);

    // Hook up data container (will be passed as the last parameter to the
    // callback handling function). 
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, httpData.get());

    // Run our HTTP GET command, capture the HTTP response code, and clean up.
    curl_easy_perform(curl);
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &httpCode);
 

    if (httpCode == 200)
    {
        std::cout << "\nGot successful response from " << url << std::endl;

        // Response looks good - done using Curl now.  Try to parse the results
        // and print them out.
        Json::Value jsonData;
        Json::Reader jsonReader;
       
        //get all the data to the JsonReader and parse it so it could be analyzed 
        if (jsonReader.parse(*httpData.get(), jsonData))
        {
          
            std::cout << "Successfully parsed JSON data" << std::endl;
            std::cout << "\nJSON data received:" << std::endl;
           
            
            auto etriesArray = jsonData["photos"];

           
            
            do
            {
                //Print the 10  first pictures.
                auto firstelement = etriesArray[i];
                std::cout << "Picture number" << i +1<< firstelement << "\n";
                i++;
            } while (i <= 9);
        
        ///////////////////////////////////////////////////////////////Get the Earth date //////////////////////////////////////////////////////////
           
            auto earthday = etriesArray[0]["earth_date"];
            std::cout << "date" << 1 << earthday << "\n";

            Json::StreamWriterBuilder builder;
            builder["earth_date"] = "";
            //earthdate = Json::writeString(builder, earthday);

            std::cout << "date11111111111111" << 1 << earthdate << "\n";


        }
        else
        {
            //Error messages
            std::cout << "Could not parse HTTP data as JSON" << std::endl;
            std::cout << "HTTP data was:\n" << *httpData.get() << std::endl;
            return 1;
        }
        
    }
    else
    {
        //URL not correct.
        std::cout << "Couldn't GET from " << url << " - exiting" << std::endl;
        return 1;
    }   

 
  
    /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/

    curl_global_init(0);
    curl = curl_easy_init();

    std :: string url2 = NASA_Earth.c_str() + earthdate + "&api_key=" + My_KEY;

    curl_easy_setopt(curl, CURLOPT_URL, url2);

    //Timeout 10sec
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10);

    // Response information.
    long httpCode2(0);
    std::unique_ptr<std::string> httpData2(new std::string());
    // Hook up data handling function.
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, callback);

    // Hook up data container (will be passed as the last parameter to the
    // callback handling function). 
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, httpData2.get());

    // Run our HTTP GET command, capture the HTTP response code, and clean up.
    curl_easy_perform(curl);
    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &httpCode2);
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    if (httpCode2 == 200)
    {
        std::cout << "\nGot successful response from " << url2 << std::endl;

        // Response looks good - done using Curl now.  Try to parse the results
        // and print them out.
        Json::Value jsonData;
        Json::Reader jsonReader;

        //get all the data to the JsonReader and parse it so it could be analyzed 
        if (jsonReader.parse(*httpData2.get(), jsonData))
        {

            std::cout << "Successfully parsed JSON data" << std::endl;
            std::cout << "\nJSON data received:" << std::endl;
            auto etriesArray = jsonData["photos"];

            do
            {
                //Print the 10  first pictures of the date earth
                auto firstelement = etriesArray[i];
                std::cout << "Picture number" << i + 1 << firstelement << "\n";
                i++;
            } while (i <= 9);
        }
        else
        {
            //Error messages
            std::cout << "Could not parse HTTP data as JSON" << std::endl;
            std::cout << "HTTP data was:\n" << *httpData2.get() << std::endl;
            return 1;
        }

    }
    else
    {
        //URL not correct.
        std::cout << "Couldn't GET from " << url2 << " - exiting" << std::endl;
        std::cout << "HTTP data was:\n" << *httpData2.get() << std::endl;
        std::cout << "HTTP data was:\n" << httpCode2  << std::endl;
        return 1;
    }
    return 0;


  }

